tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php

```diff
@@ -85,4 +85,255 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
+            '2 attributes, 1 different class each, 2nd class to be kept' => [
+                'html' => '<p class="foo">hello</p><p class="bar">world</p>',
+                'classes to keep' => ['bar'],
+                'classes expected to be removed' => ['foo'],
+            ],
+            '2 attributes, same 1 class each, that class to be kept' => [
+                'html' => '<p class="foo">hello</p><p class="foo">world</p>',
+                'classes to keep' => ['foo'],
+            ],
+            '1 attribute, 2 classes, both to be kept' => [
+                'html' => '<p class="foo bar">hello</p>',
+                'classes to keep' => ['foo', 'bar'],
+            ],
+            'first class in attribute is to be removed' => [
+                'html' => '<p class="foo bar baz">hello</p>',
+                'classes to keep' => ['bar', 'baz'],
+                'classes expected to be removed' => ['foo'],
+            ],
+            'middle class in attribute is to be removed' => [
+                'html' => '<p class="foo bar baz">hello</p>',
+                'classes to keep' => ['foo', 'baz'],
+                'classes expected to be removed' => ['bar'],
+            ],
+            'last class in attribute is to be removed' => [
+                'html' => '<p class="foo bar baz">hello</p>',
+                'classes to keep' => ['foo', 'bar'],
+                'classes expected to be removed' => ['baz'],
+            ],
+        ];
+        $matchedClassesWithExtraWhitespaceData = $this->matchedClassesWithExtraWhitespaceDataProvider();
+
+        return $nonMatchedClassesData + $matchedClassesData + $matchedClassesWithExtraWhitespaceData;
+    }
+
+    /**
+     * @return (string|string[])[][]
+     */
+    public function matchedClassesWithExtraWhitespaceDataProvider()
+    {
+        return [
+            '1 attribute, 1 class with extra whitespace, that class to be kept' => [
+                'html' => '<p class=" foo ">hello</p>',
+                'classes to keep' => ['foo'],
+            ],
+            '1 attribute, 2 classes with extra whitespace, both to be kept' => [
+                'html' => '<p class=" foo  bar ">hello</p>',
+                'classes to keep' => ['foo', 'bar'],
+            ],
+            '1 attribute, 2 classes with extra whitespace, 1 to be kept' => [
+                'html' => '<p class=" foo  bar ">hello</p>',
+                'classes to keep' => ['foo'],
+                'classes expected to be removed' => ['bar'],
+            ],
+            '1 attribute, 2 classes separated by newline, both to be kept' => [
+                'html' => "<p class=\"foo\nbar\">hello</p>",
+                'classes to keep' => ['foo', 'bar'],
+            ],
+            '1 attribute, 2 classes separated by newline, 1 to be kept' => [
+                'html' => "<p class=\"foo\nbar\">hello</p>",
+                'classes to keep' => ['foo'],
+                'classes expected to be removed' => ['bar'],
```

Comment:
Instead of relying on the default values for the parameters, please let's make the third value explicit. Then the reader does not need to know the test method interface to understand the data from the data provider.

Reply:
OK